dali/kernels/imgproc/convolution/laplacian_cpu.h

```diff
@@ -0,0 +1,311 @@
+
+    auto req_dz = sobel_dz_.Setup(ctx, in_shape, window_sizes[0]);
+    auto req_dy = sobel_dy_.Setup(ctx, in_shape, window_sizes[1]);
+    auto req_dx = sobel_dx_.Setup(ctx, in_shape, window_sizes[2]);
+
+    // Calculate max scratch memory required by sub-kernels
+    sub_scratch_sizes_ = MaxScratchSize(req_dx.scratch_sizes, req_dy.scratch_sizes);
+    sub_scratch_sizes_ = MaxScratchSize(sub_scratch_sizes_, req_dz.scratch_sizes);
+    req.scratch_sizes = sub_scratch_sizes_;
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, Intermediate, ndim> &acc,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<std::array<TensorView<StorageCPU, const W, 1>, axes>, axes>& windows,
+           const std::array<float, axes>& scale, const T& transform) {
+    // Prepare the scratchpad with all the remaining memory requested by sub-kernels
+    // TODO(michalz): Get rid of this run-time memory kind dispatch
+    PreallocatedScratchpad sub_scratch;
+    for (size_t i = 0; i < sub_scratch_sizes_.size(); i++) {
+      auto sz = sub_scratch_sizes_[i];
+      auto kind_id = static_cast<mm::memory_kind_id>(i);
+      sub_scratch.allocs[i] = BumpAllocator(
+        static_cast<char*>(ctx.scratchpad->Alloc(kind_id, sz, 64)), sz);
+    }
+
+    KernelContext sub_ctx = ctx;
+    sub_ctx.scratchpad = &sub_scratch;
+
+    // Clear the scratchpad for sub-kernels to reuse memory
+    sobel_dz_.Run(sub_ctx, acc, in, windows[0], scale[0]);
+    sub_scratch.Clear();
+    sobel_dy_.Run(sub_ctx, acc, in, windows[1], scale[1]);
+    sub_scratch.Clear();
+    sobel_dx_.Run(sub_ctx, out, in, windows[2], transform);
+  }
+
+  scratch_sizes_t sub_scratch_sizes_;
+  DzKernel sobel_dz_;
+  DyKernel sobel_dy_;
+  DxKernel sobel_dx_;
+};
+
+
+/**
+ * @brief Laplacian kernel.
+ * Provides separate specialization for 1 dimensional data. For other cases
+ * there are two specializations depending on whether ``Intermediate`` is the same as ``Out``:
+ * if not, the extra intermediate buffer is used to accumulate partial derivatives.
+ *
+ * For `axes` dimensional input data, there will be computed and summed `axes` partial derivatives,
+ * one in each direction. The `i-th` partial derivative is computed by performing
+ * a `axes`-dimensional separable convolution, where the `i-th` window is responsible for actual
+ * derivative approximation and the remaining windows provide smoothing.
+ * In total, there are ``axes * axes`` windows involved in computing laplacian. Thus,
+ * the arguments that describe windows passed to the kernel methods are `axes x axes` dimensional
+ * arrays of parameters, where ``[i][j]`` parameter refers to j-th window of i-th partial
+ * derivative.
+ */
+template <typename Intermediate, typename Out, typename In, typename W, int axes, bool has_channels,
+          typename Dummy = void>
```

Comment:
Could you put a comment, why we need the `Dummy`?

Reply:
Sure, it is to have unambiguous template specialization.

Reply:
Done